extensions/amp-viewer-integration/0.1/highlight-handler.js

```diff
@@ -96,62 +112,135 @@ export class HighlightHandler {
+    for (const key in opt_params) {
+      params[key] = opt_params[key];
+    }
+    this.viewer_.sendMessage(HIGHLIGHT_STATE, params);
+  }
+
   /**
    * @param {!HighlightInfoDef} highlightInfo
-    * @private
-    */
-  initHighlight_(highlightInfo) {
-    const ampdoc = this.ampdoc_;
-    const {win} = ampdoc;
-    const sens = findSentences(win, ampdoc.getBody(), highlightInfo.sentences);
+   * @private
+   */
+  findHighlightedNodes_(highlightInfo) {
+    const {win} = this.ampdoc_;
+    const sens = findSentences(
+        win, this.ampdoc_.getBody(), highlightInfo.sentences);
     if (!sens) {
       return;
     }
-    const spans = markTextRangeList(win, sens);
-    if (spans.length <= 0) {
+    const nodes = markTextRangeList(win, sens);
+    if (!nodes || nodes.length == 0) {
+      return;
+    }
+    this.highlightedNodes_ = nodes;
+  }
+
+  /**
+   * @param {!HighlightInfoDef} highlightInfo
+   * @private
+   */
+  initHighlight_(highlightInfo) {
+    this.findHighlightedNodes_(highlightInfo);
+    if (!this.highlightedNodes_) {
+      this.sendHighlightState_('not_found');
       return;
     }
-    for (let i = 0; i < spans.length; i++) {
-      const n = spans[i];
+    const scrollTop = this.calcTopToCenterHighlightedNodes_();
+    this.sendHighlightState_('found', dict({'scroll': scrollTop}));
+
+    for (let i = 0; i < this.highlightedNodes_.length; i++) {
+      const n = this.highlightedNodes_[i];
       n['style']['backgroundColor'] = '#ff0';
       n['style']['color'] = '#333';
     }
-    this.highlightedNodes_ = spans;
-    const viewer = Services.viewerForDoc(ampdoc);
-    const visibility = viewer.getVisibilityState();
+
+    const visibility = this.viewer_.getVisibilityState();
     if (visibility == 'visible') {
-      Services.viewportForDoc(ampdoc).animateScrollIntoView(spans[0], 500);
+      this.centerHighlightedNodes_();
     } else {
+      if (scrollTop > SCROLL_ANIMATION_HIGHT_LIMIT) {
+        Services.viewportForDoc(this.ampdoc_).setScrollTop(
+            scrollTop - SCROLL_ANIMATION_HIGHT_LIMIT);
+      }
       let called = false;
-      viewer.onVisibilityChanged(() => {
+      this.viewer_.onVisibilityChanged(() => {
         // TODO(yunabe): Unregister the handler.
-        if (called || viewer.getVisibilityState() != 'visible') {
+        if (called || this.viewer_.getVisibilityState() != 'visible') {
           return;
         }
-        Services.viewportForDoc(ampdoc).animateScrollIntoView(spans[0], 500);
-        viewer.sendMessage('highlightShown', null);
+        this.centerHighlightedNodes_();
         called = true;
       });
     }
+    listenOnce(this.ampdoc_.getBody(), 'click',
+        this.dismissHighlight_.bind(this));
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  calcTopToCenterHighlightedNodes_() {
+    const nodes = this.highlightedNodes_;
+    if (!nodes) {
+      return 0;
+    }
+    const viewport = Services.viewportForDoc(this.ampdoc_);
```

Comment:
There are two reasons I implemented a new logic:
1. I want to center multiple elements because discontinuous sentences can be highlighted.
2. Want to center elements when the header is shown (height = win.height - viewport.getPaddingTop).

Reply:
There are two reasons I implemented a new logic:
1. I want to center multiple elements because discontinuous sentences can be highlighted.
2. Want to center elements when the header is shown (height = win.height - viewport.getPaddingTop).

Reply:
In that case, I propose you do the following:

- Use the same math to calculate a center *rectangle*.
- Said rectangle is used to create an invisible element. `pointer-events: none` so that it won't block interaction.
- Use `scrollIntoView(invisibleElement, duration, curve, 'center')`.
- Remove invisible element afterwards.

This has two purposes:

- keep the viewport API stable.
- Aid in the AMP build-size reduction effort. We have a bundle size restriction in place, which would be triggered by adding a method to the viewport service. We can relax this limit, but measures like this help keep it smaller in the first place! :)

Reply:
I'm not sure I understand your suggestion correctly.

To implement your logic, I need to add new code to implement
- Said rectangle is used to create an invisible element with pointer-events: none so that it won't block interaction.

This is not very trivial and it will increase `amp-viewer-extension` a little bit. However, I can not remove any code from the current implementation except for the interface of `animateScrollToTop` in `viewport-impl.js`.

Which part of code do you think we can remove by your suggestion?

Reply:
Increasing the size of `amp-viewer-integeration` is fine. However, by modifying `Viewport`, the size of the main binary increases.

I'm suggesting that you revert the changes to `Viewport` and use  the method described above.

Reply:
That's true. But the additional size of `dist/v0.js` is just 7 bytes after gzip (78597 --> 78604, 107 bytes before gzip 247319 --> 247426).
Do 7 bytes (+0.01%) really matter in the core AMP binary size?

If 7 bytes really matter, I will remove default parameters from animateScrollToTop. Then, the size diff is only 1 byte after gzip (78597 --> 78598).

Do we really want to implement non-trivial logic in `amp-viewer-integration` just to save 7 bytes (1 byte w/o default parameters) in the core runtime?

Reply:
Sorry if I'm missing something, but how is it non-trivial? It seems like you could calculate <bottom> easily and left/right aren't an issue since we're only scrolling vertically. So it would only be a matter of doing something like:

```js
const sentinel = document.createElement('i-amphtml-sentinel');
setStyles(sentinel, {
  'position': absolute,
  'left: '0',
  'right: '0',
  'top': topAsUsuallyCalculated,
  'bottom': newLogicToCalculateBottom(),
});
this.getAmpDoc().getBody().appendChild(sentinel);
return viewport.animateScrollIntoView(// .. etc
```

Unfortunately we're only counting size before gzip.

Reply:
I meant I think it's less trivial to inject the sentinal node than to export `animateScrollTo` in viewerport, especially considering `setScrollTop` is already exported in viewport and the additional size to `v0.js` is limited.

But I do not have a strong preference. I reimplemented this based on your suggestion.